### ISSUE TITLE
Simplification of Token Generation in game

### ISSUE THREAD

**Issue by TaZi-loh** [Cockatrice/Cockatrice#590]
The token generation is quite cumbersome.
I suggest, to create a .xml table that saves a standard token for each card, that generates tokens. In game, if you have focus on a card and press ctrl+t the standard token corresponding to that card is created. If someone would like to create some special thing he could still press ctrl+t when not focused on a card and get the general token creation menu.

One could debate if another shortcut than ctrl+t should be used and if we should try to skip the table and just run a script on the fly that parses the token from the cardtext.

**Comment by tooomm** [comment]
Sounds nice. Even a right click on the card with "Create card related token" would be great.
But someone needs to maintain that xml table...
And since we don't want cockatrice to be mtg only it needs to be optional and not included.

This might work if someone is taking care of this like the token file from @Psithief
Maybe an addition to his file will work as well, if cockatrice checks for it.

**Comment by Daenyth** [comment]
We may be able to do it for transform and split cards but general token auto creation sounds kind of tricky.

I think our best option is to make it easier to add tokens to deck files so they're at the top of the list

**Comment by Psithief** [comment]
That can be done wholly with regular expressions. No need for me to get involved. :wink:

Here's an untested but fairly close example.
```
(\d+|X)/(\d+|X) (((white|blue|black|red|green|colorless)( and (white|blue|black|red|green)?)?(\w+) ((artifact )?(enchantment )?(creature )?tokens?)(with .+?)?(named .+?)? onto the battlefield(\. It has .+?(\n|$))?
```

**Comment by Psithief** [comment]
@Daenyth 
Originally a some-what joking reply, it only took me 20 minutes to actually make that regular expression work against `cards.xml`. Does this suddenly seem viable or is it just me?

[Example 1 (Gatherer)](http://gatherer.wizards.com/Pages/Search/Default.aspx?action=advanced&output=spoiler&method=visual&text=+[m/((\\senchantment)?(\\sartifact)?(\\screature)?\\stokens?).+onto.the.battlefield/]+[m/(\\d+|X)/(\\d+|X)(\\s(white|blue|black|red|green|colorless)(\\sand\\s(white|blue|black|red|green)?)?)?/])

**Comment by TaZi-loh** [comment]
yeah regular expression was what I thought about when I edited the starting post. I don't know how to apply a regular expression in c++ though atm. All the other stuff would be easily codeable. This would just be a slightly modified instance of the ctrl+j command.

**Comment by Psithief** [comment]
I don't think it entirely makes sense to actually do it in Cockatrice unless it was also to support custom card sets.

The alternative is just expanding the Oracle application to download some additional data files that have all of this card -> token information.

**Comment by ZeldaZach** [comment]
Relating to @Psithief's statement... Why don't we have oracle download the tokens file from his github page in addition to all the sets? Would be helpful

**Comment by tooomm** [comment]
@ZeldaZach I would love to see that too! :+1: 
I don't know how many people I already explained where these nice tokens come from and how to install them...

@Psithief has done some nice work. Let's spread his gorgeous tokens. :)

**Comment by Psithief** [comment]
Please direct any ideas to #473. :+1:

**Comment by TaZi-loh** [comment]
ok, to summarize the diskussion we need to
* write a small script applying a regular expression to extract the information which card generates which token from the cards.xml to some token_generation.xml
* change the oracle such that it is able to download the tokens.xml and the token_generation.xml
* add a shortcut that creates an appropriate token, when a card is focused and the card name is contained in token_generation.xml

is everybody ok with this?

**Comment by Daenyth** [comment]
Well there's two parts to this. Detecting related card names and creating tokens for related cards. We can address the cockatrice code and leave the rest for later.

**Comment by tooomm** [comment]
@TaZi-loh 
Related to your 2nd point: Why do we need to download the token_generation.xml at all?
Wouldn't it be much easier if oracle can create it itself out of the cards.xml with the script (or directly out of the mtgjson) while running anyway or even store this information inside cards.xml too?

Token file downloading in oracle would be awesome...

**Comment by TaZi-loh** [comment]
I think it would not be a "clean" solution to put the information into the cards.xml but in a seperate file. By changing the oracle I meant, that there should be a checkbox like "downloading tokens" which would cause the tokens.xml and token_generation.xml to be downloaded. Wether the tokens_genration.xml would be generated on the fly by the oracle code, or by a script run manually before is still to be determined, but I would love a suggestion for that.

**Comment by tooomm** [comment]
Store this information in a seperate file is totally fine!

But I still can't follow you completely.
You are talking about "downloading tokens which would cause the [...] token_generation.xml to be downloaded". But then you speak about "generated on the fly by oracle code, or by a script run manually before".
So do you want to provide a *download* for the token_generation.xml or do you want to *generate* it?

I'm not a fan of the download option because in this case that file needs to be maintained and provided seperatly by someone. I don't think that's needed and useful here.
But that depends on the generation with the script running without problems.

**Comment by Psithief** [comment]
We don't parse gatherer in Oracle either, we download mtgjson's prepared version.

Including code to generate that file in Oracle ties its functionality to Cockatrice's release cycle. Avoiding that scenario is worth consideration.

**Comment by Psithief** [comment]
To load the right art for a token generating card, I would need to include a unique id for each token variant. Is anyone familiar with hashing functions?

**Comment by ctrlaltca** [comment]
By now cockatrice uses the card name as the card identifier, so different tokens with the same name (eg Goblin) are using a white space suffix as a workaround. I think we should solve this issue, too.

**Comment by Daenyth** [comment]
A unique hash makes sense, taking into account name/type line/pt/abilities.
If we extend the "create related card" menu to be able to do tokens, that
would be a nice UI solution

**Comment by Daenyth** [comment]
related to #1203

**Comment by ZeldaZach** [comment]
Fixed in #1722
